src/proof/lfsc/lfsc_printer.cpp

```diff
@@ -0,0 +1,515 @@
+          // print the identifier
+          out->printNode(pidNode);
+          // Print the body of the proof with a fresh proof letification. We can
+          // keep the assumption map and the let binding (for terms).
+          std::vector<const ProofNode*> pletListNested;
+          std::map<const ProofNode*, size_t> pletMapNested;
+          const ProofNode* curBody = cur->getChildren()[0].get();
+          computeProofLetification(curBody, pletListNested, pletMapNested);
+          printProofLetify(
+              out, curBody, lbind, pletListNested, pletMapNested, passumeMap);
+          // print ")"
+          out->printCloseRule();
+        }
+        else
+        {
+          // assert that we should traverse cur when letifying
+          Assert(d_lpltc.shouldTraverse(cur));
+          // a normal rule application, compute the proof arguments, which
+          // notice in the case of PI also may modify our passumeMap.
+          std::vector<PExpr> args;
+          if (computeProofArgs(cur, args))
+          {
+            processingChildren.insert(cur);
+            // will revisit this proof node to close parentheses
+            visit.push_back(PExpr(cur));
+            std::reverse(args.begin(), args.end());
+            visit.insert(visit.end(), args.begin(), args.end());
+            // print the rule name
+            out->printOpenRule(cur);
+          }
+          else
+          {
+            // could not print the rule, trust for now
+            Node res = d_tproc.convert(cur->getResult());
+            res = lbind.convert(res, "__t", true);
+            out->printTrust(res, r);
+            if (d_trustWarned.find(r) == d_trustWarned.end())
+            {
+              d_trustWarned.insert(r);
+            }
+          }
+        }
+      }
+      else
+      {
+        processingChildren.erase(cur);
+        out->printCloseRule();
+      }
+    }
+    // case 2: printing a node
+    else if (!curn.isNull())
+    {
+      // it has already been converted to internal form, we letify it here
+      Node curni = lbind.convert(curn, "__t", true);
+      out->printNode(curni);
+    }
+    // case 3: printing a type node
+    else if (!curtn.isNull())
+    {
+      out->printTypeNode(curtn);
+    }
+    // case 4: a hole
+    else
+    {
+      out->printHole();
+    }
+  } while (!visit.empty());
+}
+
+bool LfscPrinter::computeProofArgs(const ProofNode* pn,
+                                   std::vector<PExpr>& pargs)
+{
+  const std::vector<std::shared_ptr<ProofNode>>& children = pn->getChildren();
+  std::vector<const ProofNode*> cs;
+  for (const std::shared_ptr<ProofNode>& c : children)
+  {
+    cs.push_back(c.get());
+  }
+  PfRule r = pn->getRule();
+  const std::vector<Node>& args = pn->getArguments();
+  std::vector<Node> as;
+  for (const Node& a : args)
+  {
+    Node ac = d_tproc.convert(a);
+    Assert(!ac.isNull());
+    as.push_back(ac);
+  }
+  PExprStream pf(pargs, d_tt, d_ff);
```

Comment:
Can you please generally document this block? Just to clarify its purpose.

Reply:
Done